src/macros.jl

```diff
@@ -359,23 +368,42 @@ macro wrap(ex)
 
     fullargs = ex.args[2:length(ex.args)]
 
-    # strip type annotation from arguments for use in the call to GAP
-    args = [x isa Symbol ? x : x.args[1] for x in fullargs]
-
+    # splits the arguments with type annotations into expressions for the lhs and rhs
+    # of the call of the form `func(args) = GAP.Globals.func(args)`:
+    # - type annotations have to be removed for the rhs
+    # - if the type annotation is `GapObj`, the rhs has to be wrapped in `GAP.GapObj(...)` and the lhs has no type annotation
+    tempargs = [
+        begin
+            if x isa Symbol
+                (x, x)
+            elseif x.head == :(::) && length(x.args) == 2
+                var = x.args[1]
+                typeannot = x.args[2]
+                if typeannot in [:GapObj, :(GAP.GapObj)]
+                    (var, :(GAP.GapObj($var)))
```

Comment:
Do all of the constructors have a (inferrable) contract to return objects of the specific type?
One could adapt this here to 
```suggestion
                    (var, :(GAP.GapObj($var)::GAP.GapObj))
```
to make sure

Reply:
I am not sure what you mean with "inferrable contract", I don't think such a thing exists in Julia? In the end it's the same as with every such thing in Julia: Even `String(x)` may not have an inferable type if some package (loaded outside of your control) decides to add a crazy method to it.

Reply:
That's exactly what I was talking about. Someone (maybe even GAP.jl itself) could add a method to GapObj that does not return a GapObj. With a type annotation as suggested above, we could make sure that we indeed get a GapObj in the coercion

Reply:
Fair enough, so yeah add that :-)